Add OpenAI embeddings instrumentation

drewby opened the pull request:
# Description

This PR adds instrumentation for OpenAI's embeddings API in the GenAI instrumentation suite. The implementation follows the OpenTelemetry semantic conventions for generative AI systems and provides automatic instrumentation for the OpenAI Python client when using embeddings functionality.

The implementation captures important metadata about embedding operations including model, dimensions, and relevant timing information while respecting sensitive data handling practices.

* Added instrumentation for both synchronous and asynchronous OpenAI embedding API calls
* Implemented span and metrics using existing attributes, with two new custom:
  * `ai.embedding.dimensions` - Number of dimensions in the embedding vectors
  * `ai.embedding.encoding_format` - The encoding format of the embedding vectors response (base64 or float)
* Capturing input text content (disabled by default for privacy)
* Added a usage example called `embeddings`

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [x] Unit tests using mock responses to verify proper span creation and attribute population
- [x] Integration tests with the OpenAI client against a mock server
- [x] Manual testing using examples/embeddings with real OpenAI service

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

lmolkova commented on instrumentation-genai/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/patch.py:
```diff
@@ -262,6 +406,91 @@ def _set_response_attributes(
+        set_span_attribute(
+            span,
+            GenAIAttributes.GEN_AI_USAGE_OUTPUT_TOKENS,
+            result.usage.total_tokens,
```
total is not output, right? we should not have any output tokens for embeddings if I recall correctly

xrmx replied:
Yes, we have only input tokens in semconv

lmolkova commented on instrumentation-genai/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/patch.py:
```diff
@@ -262,6 +406,91 @@ def _set_response_attributes(
+        # Emit input event
+        input_event_attributes = {
+            GenAIAttributes.GEN_AI_SYSTEM: GenAIAttributes.GenAiSystemValues.OPENAI.value,
+            EventAttributes.EVENT_NAME: "gen_ai.embeddings.input",
```
I'd postpone defining any new events until we have clarity on https://github.com/open-telemetry/semantic-conventions/issues/2010

Also why not `gen_ai.user.message`?

drewby replied:
Recording inputs/outputs is a fundamental feature we discussed, so I think we need to record as something. The current implementation for chat completions is events, so that would be the most consistent until we decide on 2010. If the decision is to move to attributes, we are going to need to do breaking change PR anyway.

I could reuse the existing event name, but its not really a `user message` in the same way as text completions.

lmolkova replied:
Let's sync and check if we can do it in a separate PR as a follow up leaving this one to non-controversial stuff

drewby replied:
So I remove inputs/outputs from this PR and then we can add it as another PR with a more focused discussion? That works, just want to make sure I understand.

xrmx replied:
@drewby yes, I think dropping these events would be helpful to get the rest in

lmolkova commented on instrumentation-genai/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/patch.py:
```diff
@@ -262,6 +406,91 @@ def _set_response_attributes(
+
+            output_event_attributes = {
+                GenAIAttributes.GEN_AI_SYSTEM: GenAIAttributes.GenAiSystemValues.OPENAI.value,
+                EventAttributes.EVENT_NAME: "gen_ai.embeddings.output",
```
see my prev comment, I'd rather not define any new events for now

lmolkova commented on instrumentation-genai/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/patch.py:
```diff
@@ -149,6 +150,137 @@ async def traced_method(wrapped, instance, args, kwargs):
+            name=span_name,
+            kind=SpanKind.CLIENT,
+            attributes=span_attributes,
+            end_on_exit=False,
```
why not close it on exit and remove line 268 (`span.end`) ? there is no streaming in this case

lmolkova commented on instrumentation-genai/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/patch.py:
```diff
@@ -160,15 +292,22 @@ def _record_metrics(
         ],
     }
 
+    if "gen_ai.embeddings.dimensions" in span_attributes:
```
we can't use span attributes to record metrics - spans are sampled and won't have any attributes when sampled out.

drewby replied:
The parameter is called span_attributes, but it is from the previous PR. 

Its just a dictionary created during the request by `chat_completions_create` and now `embeddings_create`, so not subject to sampling. I renamed it here in _record_metrics to just `request_attributes` for clarity. If needed, I can rename it in the create methods for better readability as well.

lmolkova commented on instrumentation-genai/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/patch.py:
```diff
@@ -210,13 +349,18 @@ def _record_metrics(
 
-        completion_attributes = {
+        token_count = (
+            result.usage.total_tokens
```
should we report output tokens for embeddings?

lmolkova commented on instrumentation-genai/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/patch.py:
```diff
@@ -262,6 +406,91 @@ def _set_response_attributes(
+
+    # Emit events for embeddings if content capture is enabled
+    if capture_content:
+        from opentelemetry.semconv._incubating.attributes import (
+            event_attributes as EventAttributes,
+        )
```
do we need to import conditionally?

lmolkova reviewed (commented):
thank you!!!

xrmx commented on instrumentation-genai/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py:
```diff
@@ -192,38 +192,66 @@ def get_llm_request_attributes(
+        # Add encoding format if specified
+        # Using a custom attribute "gen_ai.embeddings.encoding_format". Will propose to semantic conventions.
+        if "encoding_format" in kwargs:
+            attributes["gen_ai.embeddings.encoding_format"] = kwargs[
```
Why not using `gen_ai.request.encoding_formats` ?

xrmx commented on instrumentation-genai/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/patch.py:
```diff
@@ -149,6 +150,137 @@ async def traced_method(wrapped, instance, args, kwargs):
+
+        # Using a custom attribute "gen_ai.embeddings.dimensions". Will propose to semantic conventions.
+        if "dimensions" in kwargs and kwargs["dimensions"] is not None:
+            span_attributes["gen_ai.embeddings.dimensions"] = kwargs[
```
Don't forget to propose this to semconv

xrmx commented on instrumentation-genai/opentelemetry-instrumentation-openai-v2/examples/embeddings/requirements.txt:
```diff
@@ -0,0 +1,6 @@
+openai~=1.57.3
+
+opentelemetry-sdk~=1.30.0
```
You'll need to bump these

xrmx commented on instrumentation-genai/opentelemetry-instrumentation-openai-v2/tests/test_async_embeddings.py:
```diff
@@ -0,0 +1,311 @@
+    else:
+        assert GenAIAttributes.GEN_AI_USAGE_INPUT_TOKENS not in span.attributes
+
+    if output_tokens:
```
leftover? also for sync

xrmx commented on instrumentation-genai/opentelemetry-instrumentation-openai-v2/tests/test_async_embeddings.py:
```diff
@@ -0,0 +1,311 @@
+            GenAIAttributes.GEN_AI_USAGE_OUTPUT_TOKENS not in span.attributes
+        )
+
+    assert server_address == span.attributes[ServerAttributes.SERVER_ADDRESS]
```
I think we could also assert server port and gen_ai.request.encoding_formats here

xrmx commented on instrumentation-genai/opentelemetry-instrumentation-openai-v2/tests/test_embeddings.py:
```diff
@@ -0,0 +1,444 @@
+
+    # Verify dimensions attribute is set correctly
+    assert (
+        spans[0].attributes["gen_ai.embeddings.dimension.count"] == dimensions
```
This is still not specified in semantic conventions right?

drewby replied:
Correct, I'll work on a PR to the semantic conventions.

xrmx commented on instrumentation-genai/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py:
```diff
@@ -192,38 +192,67 @@ def get_llm_request_attributes(
+
+        # Add encoding format if specified
+        if "encoding_format" in kwargs:
+            attributes["gen_ai.request.encoding_formats"] = kwargs[
```
This should be a list of encoding_formats, not a single string . On the assert side they would become tuples

anuraaga commented:
Hey - just wondering what's the status of this PR as interested in having embeddings instrumented here. If there's anything I can do to help on it, let me know.

drewby commented:
@xrmx @lmolkova I've made several updates based on the feedback. I hope I addressed most concerns. There is one check not passing but seems unrelated to openai library.

xrmx commented on instrumentation-genai/opentelemetry-instrumentation-openai-v2/examples/embeddings/requirements.txt:
```diff
@@ -0,0 +1,6 @@
+opentelemetry-sdk~=1.36.0
+opentelemetry-exporter-otlp-proto-grpc~=1.36.0
+opentelemetry-distro~=0.57b0
+opentelemetry-instrumentation-openai-v2~=2.1b0
```
```suggestion
opentelemetry-instrumentation-openai-v2~=2.2b0
```
